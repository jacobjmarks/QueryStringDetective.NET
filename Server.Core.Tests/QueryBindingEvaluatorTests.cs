namespace Server.Core.Tests;

public class QueryBindingEvaluatorTests
{
    public static readonly TheoryData<string> TestCases = new()
    {
        "c",
        "foo",
        "a&q=b&q=c",
        "null",
        "0",
        "1",
        "-1",
        "1&q=2&q=3",
        "1&q=2&q=.3",
        "-1&q=0&q=1",
        $"{sbyte.MinValue}",
        $"{sbyte.MaxValue}",
        $"{byte.MaxValue}",
        $"{short.MinValue}",
        $"{short.MaxValue}",
        $"{ushort.MaxValue}",
        $"{int.MinValue}",
        $"{int.MaxValue}",
        $"{long.MinValue}",
        $"{long.MaxValue}",
        $"{ulong.MaxValue}",
        Uri.EscapeDataString($"{double.MinValue}"),
        Uri.EscapeDataString($"{double.MaxValue}"),
        $"{decimal.MinValue}",
        $"{decimal.MaxValue}",
        Uri.EscapeDataString($"{float.MinValue}"),
        Uri.EscapeDataString($"{float.MaxValue}"),
    };

    [Theory]
    [MemberData(nameof(TestCases))]
    public async Task AssertBinding(string queryString)
    {
        using var bindingEvaluator = new QueryBindingEvaluator();
        var results = await bindingEvaluator.EvaluateAsync("?q=" + queryString);

        var content = "Input: " + queryString
            + "\n\nBindings:\n" + string.Join("\n", results.Where(r => !r.IsErroneous)
                    .Select(r => $"{r.Type,-10} {r.Result}"))
            + "\n\nErroneous:\n" + string.Join("\n", results.Where(r => r.IsErroneous)
                    .Select(r => $"{r.Type,-10} {r.Error!.Message}"));

        content.Should().MatchSnapshot(SnapshotNameExtension.Create(ToValidFileName(queryString)));
    }

    private static string ToValidFileName(string value)
    {
        foreach (var invalidChar in Path.GetInvalidFileNameChars())
            value = value.Replace(invalidChar, '_');

        return value;
    }
}