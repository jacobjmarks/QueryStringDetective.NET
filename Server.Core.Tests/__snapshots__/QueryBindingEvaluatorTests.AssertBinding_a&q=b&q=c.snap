Input: a&q=b&q=c

Bindings:

string
  MinimalApis
    r = "a,b,c"
  Controllers
    r = "a"

string?
  MinimalApis
    r = "a,b,c"
  Controllers
    r = "a"

string[]
  MinimalApis
    r = ["a","b","c"]
  Controllers
    r = ["a","b","c"]

char
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "char q" from "a,b,c".)
  Controllers
    r = "a"

char?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<char> q" from "a,b,c".)
  Controllers
    r = "a"

char[]
  MinimalApis
    r = ["a","b","c"]
  Controllers
    r = ["a","b","c"]

int
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "int q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

int?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<int> q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

int[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "int[] q" from "a".)
  Controllers
    e = 400 Bad Request (The value 'a' is not valid.)

double
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "double q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

double?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<double> q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

double[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "double[] q" from "a".)
  Controllers
    e = 400 Bad Request (The value 'a' is not valid.)

float
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "float q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

float?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<float> q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

float[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "float[] q" from "a".)
  Controllers
    e = 400 Bad Request (The value 'a' is not valid.)

decimal
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "decimal q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

decimal?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<decimal> q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

decimal[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "decimal[] q" from "a".)
  Controllers
    e = 400 Bad Request (The value 'a' is not valid.)

bool
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "bool q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

bool?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<bool> q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

bool[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "bool[] q" from "a".)
  Controllers
    e = 400 Bad Request (The value 'a' is not valid.)

uint
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "uint q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

uint?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<uint> q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

uint[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "uint[] q" from "a".)
  Controllers
    e = 400 Bad Request (The value 'a' is not valid.)

long
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "long q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

long?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<long> q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

long[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "long[] q" from "a".)
  Controllers
    e = 400 Bad Request (The value 'a' is not valid.)

ulong
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "ulong q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

ulong?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<ulong> q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

ulong[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "ulong[] q" from "a".)
  Controllers
    e = 400 Bad Request (The value 'a' is not valid.)

sbyte
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "sbyte q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

sbyte?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<sbyte> q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

sbyte[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "sbyte[] q" from "a".)
  Controllers
    e = 400 Bad Request (The value 'a' is not valid.)

short
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "short q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

short?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<short> q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

short[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "short[] q" from "a".)
  Controllers
    e = 400 Bad Request (The value 'a' is not valid.)

ushort
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "ushort q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

ushort?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<ushort> q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

ushort[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "ushort[] q" from "a".)
  Controllers
    e = 400 Bad Request (The value 'a' is not valid.)

byte
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "byte q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

byte?
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "Nullable<byte> q" from "a,b,c".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)

byte[]
  MinimalApis
    e = 400 Bad Request (Failed to bind parameter "byte[] q" from "a".)
  Controllers
    e = 400 Bad Request (The value 'a,b,c' is not valid.)
